src/ngx_http_lua_kong_log.c

```diff
@@ -14,45 +14,150 @@
+    (NGX_ERROR > INT32_MAX) ||     \
+    (NGX_ERROR < INT32_MIN)
+#error "The int32_t type is not suitable for NGX_OK and NGX_ERROR"
+#endif
+
+static ngx_int_t g_ready_for_dynamic_log_level = 0;
+
+void
+ngx_http_lua_kong_log_set_ready_for_dynamic_log_level(ngx_int_t is_ready)
+{
+    /*
+    ** This function will be called when the following functions
+    ** are about to start event loop:
+    **   * ngx_worker_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_single_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_cache_manager_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_cache_loader_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_privileged_agent_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **
+    ** This function will be called immediately when the event loop has just been stopped:
+    **   * ngx_worker_process_exit in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_master_process_exit in the file src/os/unix/ngx_process_cycle.c.
+    **
+    ** Once the event loop is started,
+    ** we can confirm the data structure and other resources are ready for
+    ** the dynamic log level feature.
+    **
+    ** Please see the function ngx_worker_process_cycle in the file
+    ** src/os/unix/ngx_process_cycle.c.
+    */
+    g_ready_for_dynamic_log_level = is_ready;
+}
 
-int
-ngx_http_lua_kong_ffi_set_log_level(ngx_http_request_t *r, int level)
+int32_t
+ngx_http_lua_kong_ffi_set_dynamic_log_level(uint32_t log_level, uint32_t timeout)
 {
-    ngx_log_t                   *log;
-    ngx_listening_t             *ls;
-    ngx_http_core_loc_conf_t    *clcf;
-    ngx_uint_t                   i;
+    ngx_http_lua_kong_main_conf_t *lkmcf =
+        ngx_http_cycle_get_module_main_conf(ngx_cycle, ngx_http_lua_kong_module);
 
-    if (r == NULL) {
-        return NGX_ERROR;
-    }
+    ngx_http_lua_kong_assert(lkmcf != NULL);
 
-    clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
-    clcf->error_log->log_level = (ngx_uint_t) level;
+    if (timeout == 0) {
+        /*
+        ** if the timeout is 0,
+        ** we will disable the dynamic log level,
+        ** that is, the log_level will be restored to the default setting from log_level in kong.conf.
+        ** and ignore the first parameter.
+        */
+        lkmcf->dynamic_log_level_set = 0;
+        return NGX_OK;
+    }
 
-    if (ngx_cycle == NULL) {
-        return NGX_ERROR;
+    /*
+    ** if the log level is invalid,
+    ** we will not change the current log level,
+    ** and return an error.
+    */
+    switch (log_level)
+    {
+        case NGX_LOG_STDERR: case NGX_LOG_EMERG:
+        case NGX_LOG_ALERT:  case NGX_LOG_CRIT:
+        case NGX_LOG_ERR:    case NGX_LOG_WARN:
+        case NGX_LOG_NOTICE: case NGX_LOG_INFO:
+        case NGX_LOG_DEBUG:
+            break;
+        default:
+            return NGX_ERROR;
     }
```

Comment:
Could be `if (level > NGX_LOG_DEBUG || level < NGX_LOG_STDERR)`.
See `ngx_http_lua_ffi_errlog_set_filter_level`

Reply:
`NGX_LOG_STDERR` is `0`, `NGX_LOG_DEBUG` is `8`, all of them are integers.

Reply:
done